include/aws/crt/mqtt/Mqtt5Listener.h

```diff
@@ -0,0 +1,344 @@
+
+                /**
+                 * Underlying c classes for data storage and operational members
+                 */
+                aws_mqtt5_listener *m_listener;
+
+                std::map<Crt::String, Crt::Mqtt5::OnPublishReceivedHandler> m_subscriptionMap;
+
+                Allocator *m_allocator;
+            };
+
+            /**
+             * An MQTT5 listener.
+             *
+             * This is a move-only type. Unless otherwise specified, all function arguments need only to live through
+             * the duration of the function call.
+             */
+            class AWS_CRT_CPP_API Mqtt5Listener final : public std::enable_shared_from_this<Mqtt5Listener>
+            {
+              public:
+                /**
+                 * Factory function for mqtt5 listener
+                 *
+                 * @param options: Mqtt5ListenerOption
+                 * @param client: Mqtt5Client to listen to
+                 * @param allocator allocator to use
+                 * @return a new mqtt5 listener
+                 */
+                static std::shared_ptr<Mqtt5Listener> NewMqtt5Listener(
+                    const Mqtt5ListenerOptions &options,
+                    const std::shared_ptr<Mqtt5Client> &client,
+                    Allocator *allocator = ApiAllocator()) noexcept;
+
+                /**
+                 * Get shared pointer of the Mqtt5ListenerCore. Mqtt5ListenerCore is inherited to
+                 * enable_shared_from_this to help with memory safety.
+                 *
+                 * @return shared_ptr for the Mqtt5ListenerCore
+                 */
+                std::shared_ptr<Mqtt5Listener> getptr() { return shared_from_this(); }
+
+                /**
+                 * @return true if the instance is in a valid state, false otherwise.
+                 */
+                operator bool() const noexcept { return m_listener_core != nullptr; };
+
+                /**
+                 * @return the value of the last aws error encountered by operations on this instance.
+                 */
+                int LastError() const noexcept { return aws_last_error(); };
+
+                /*
+                 * Insert the topic and callback in the map
+                 */
+                void Subscribe(Crt::String topic, Crt::Mqtt5::OnPublishReceivedHandler callback) noexcept;
+
+                /*
+                 * Remove the topic and callback in the map
+                 */
+                void Unsubscribe(Crt::String) noexcept;
+
+                virtual ~Mqtt5Listener();
+
+              private:
+                Mqtt5Listener(
+                    const Mqtt5ListenerOptions &options,
+                    const std::shared_ptr<Mqtt5Client> &client,
+                    Allocator *allocator) noexcept;
+
+                /*
+                 * Reference to listener core.
+                 */
+                std::shared_ptr<Mqtt5ListenerCore> m_listener_core;
```

Comment:
unique pointer

Reply:
The `Mqtt5ListenerCore` need to be a shared_ptr, as the `Mqtt5ListenerCore` has a self-reference shared pointer.